File: sdk-source/cpp/examples/quickstart.cpp
// Runnable quickstart for the Aegis C++ SDK.
//
//   cmake -B build && cmake --build build
//   AEGIS_APP_KEY=... ./build/aegis_quickstart
#include <cstdlib>
#include <iostream>

#include "aegis/aegis.hpp"

static std::string envOr(const char* key, const std::string& fallback) {
    const char* value = std::getenv(key);
    return (value && *value) ? std::string(value) : fallback;
}

int main() {
    aegis::Options options;
    options.baseUrl = envOr("AEGIS_BASE_URL", "http://localhost:8080");
    options.appKey = envOr("AEGIS_APP_KEY", "");
    options.version = "1.0.0";

    try {
        aegis::Client client(options);

        auto info = client.init();
        std::cout << "initialized: " << info["status"].asString() << "\n";

        if (info["version"]["update_required"].asBool()) {
            std::cout << "mandatory update: " << info["version"]["latest"].asString() << "\n";
            return 0;
        }

        auto auth = client.login(envOr("AEGIS_USERNAME", "demo"), envOr("AEGIS_PASSWORD", "demo-password"));
        std::cout << "signed in as " << auth["user"]["username"].asString() << "\n";

        client.setVariable("last_seen", "now");
        std::cout << "variables: " << client.getVariables("user").dump() << "\n";
        std::cout << "downloads: " << client.downloads().dump() << "\n";

        client.startHeartbeat(std::chrono::seconds(60), [](const std::string& reason) {
            std::cout << "session revoked: " << reason << "\n";
        });

        client.logout();
        std::cout << "signed out.\n";
        return 0;
    } catch (const aegis::AegisError& error) {
        std::cerr << "Aegis error [" << error.code() << "] " << error.what() << "\n";
        return 1;
    }
}

File: sdk-source/rust/examples/quickstart.rs
//! Runnable quickstart for the Aegis Rust SDK.
//!
//! AEGIS_BASE_URL=https://your-aegis-host AEGIS_APP_KEY=pk_live_... cargo run --example quickstart
use aegis_sdk::{Aegis, AegisOptions, Scope};
use std::time::Duration;

#[tokio::main]
async fn main() {
    let base_url = std::env::var("AEGIS_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".into());
    let app_key = std::env::var("AEGIS_APP_KEY").unwrap_or_default();

    let client = match Aegis::new(AegisOptions::new(base_url, app_key).version("1.0.0")) {
        Ok(client) => client,
        Err(error) => return eprintln!("{error}"),
    };

    let info = match client.init().await {
        Ok(info) => info,
        Err(error) => return eprintln!("aegis error [{}] {}", error.code, error.message),
    };
    println!("initialized: {:?}", info.status);

    if info.version.as_ref().map(|v| v.update_required).unwrap_or(false) {
        println!("mandatory update available");
        return;
    }

    let username = std::env::var("AEGIS_USERNAME").unwrap_or_else(|_| "demo".into());
    let password = std::env::var("AEGIS_PASSWORD").unwrap_or_else(|_| "demo-password".into());

    match client.login(&username, &password).await {
        Ok(auth) => println!("signed in as {}", auth.user.username),
        Err(error) => return eprintln!("aegis error [{}] {}", error.code, error.message),
    }

    if let Ok(vars) = client.get_variables(Scope::User, None).await {
        println!("user variables: {:?}", vars.variables);
    }
    let _ = client.set_variable("last_seen", "now", Scope::User, None).await;

    let beat = client.start_heartbeat(Duration::from_secs(60), |reason| {
        eprintln!("session revoked: {reason}");
    });
    tokio::time::sleep(Duration::from_secs(2)).await;
    beat.abort();

    let _ = client.logout().await;
    println!("signed out.");
}